app/components/primer/auto_complete.rb

```diff
@@ -1,88 +1,116 @@
   # @accessibility
-  #   Always provide a label for the `Autocomplete` component.
+  #   Always provide an accessible label to help the user interact with the input element and listbox popup.
+  #
+  #   To show a visible label, set the `label` slot.
+  #
+  #   If you do not wish to provide a visible label, you must set an `aria-label` attribute. You may set
+  #   `:"aria-label"` directly on `AutoComplete` instead of the slots and Primer will apply it to the correct elements.
   class AutoComplete < Primer::Component
     status :beta
 
     DEFAULT_INPUT_TYPE = :text
     INPUT_TYPE_OPTIONS = [DEFAULT_INPUT_TYPE, :search].freeze
 
+    # Optionally render a visible label. See <%= link_to_accessibility %>
+    #
+    # @param for [Symbol] id of input
+    # @param system_arguments [Hash] <%= link_to_system_arguments_docs %>
+    renders_one :label, lambda { |**system_arguments|
+      raise ArgumentError, "missing keyword: for" if !system_arguments.key?(:for) && !Rails.env.production?
+
+      @input_id_for_label = system_arguments[:for]
+      system_arguments[:tag] = :label
+      Primer::BaseComponent.new(**system_arguments)
+    }
+
     # Required input used to search for results
     #
     # @param type [Symbol] <%= one_of(Primer::AutoComplete::INPUT_TYPE_OPTIONS) %>
     # @param system_arguments [Hash] <%= link_to_system_arguments_docs %>
     renders_one :input, lambda { |type: DEFAULT_INPUT_TYPE, classes: "form-control", **system_arguments|
       system_arguments[:tag] = :input
+      system_arguments[:role] = :combobox
+
+      aria_label = system_arguments[:"aria-label"] || system_arguments.dig(:aria, :label) || @aria_label
+      if @input_id_for_label.present? && @input_id_for_label == system_arguments[:id]
+      elsif aria_label.present?
+        system_arguments[:"aria-label"] = aria_label
+      else
+        raise ArgumentError, "Accessible label is required." unless Rails.env.production?
+      end
+
       system_arguments[:type] = fetch_or_fallback(INPUT_TYPE_OPTIONS, type, DEFAULT_INPUT_TYPE)
       system_arguments[:classes] = classes
       Primer::BaseComponent.new(**system_arguments)
     }
 
     # Optional icon to be rendered before the input. Has the same arguments as <%= link_to_component(Primer::OcticonComponent) %>.
+    #
     renders_one :icon, Primer::OcticonComponent
 
     # Customizable results list.
     #
     # @param system_arguments [Hash] <%= link_to_system_arguments_docs %>
     renders_one :results, lambda { |**system_arguments|
       system_arguments[:tag] = :ul
-      system_arguments[:id] = @id
+      system_arguments[:role] = :listbox
+      system_arguments[:id] = @list_id
       system_arguments[:classes] = class_names(
         "autocomplete-results",
         system_arguments[:classes]
       )
 
+      aria_label = system_arguments[:"aria-label"] || system_arguments.dig(:aria, :label) || @aria_label
+      system_arguments[:"aria-label"] = aria_label if aria_label.present?
+
```

Comment:
The order of precedence for labeling of `<ul role="listbox">` will be:

- Setting `aria-label` passed to this slot
- Setting `aria-label` passed to parent

For now, we won't raise if this listbox is missing a label (API is getting complex and does not seem *as critical*), but happy to revisit this.